### ISSUE TITLE
collectionAssertion for the new infix API

### ISSUE THREAD

**Issue by robstoll** [robstoll/atrium#227]
*Platform* (jvm, js, android): jvm
*Extension* (none, kotlin 1.3, jdk8): none

## Code related feature

We want to migrate the `api-cc-infix-en_GB...` modules which are based on `Assert` to `Expect`. We have already created the corresponding
modules `api-infix-en_GB...`. Also there are already `fluent-en_GB...` modules where we have already done the transition from `Assert` to `Expect`

Following the things you need to do:

- copy the file `collectionAssertions.kt` from atrium-api-fluent-en_GB-common to atrium-api-infix-en_GB-common
  - adjust the signatures of the assertion functions so that they are infix functions (see collectionAssertions.kt in atrium-api-cc-infix-en_GB-common for a guideline)
  - convert `isEmpty()` to `toBe Empty` (see collectionAssertions.kt in atrium-api-cc-infix-en_GB-common for a guideline)
  - convert `isNotEmpty()` to `notToBe Empty`
- copy the `CollectionAssertionsSpec.kt` from atrium-api-fluent-en_GB-common to atrium-api-infix-en_GB-common
  - introduce a function in the companion object for each assertion function so that we see the infix API in action (see CollectionAssertionsSpec in atrium-api-cc-infix-en_GB-jvm but use fun1, fun2 etc. to retrieve the name of the function instead of `KFunction<...>` -- use `fun1<..>(...).name`)
- copy the `CollectionFeatureAssertionsSpec.kt` from atrium-api-fluent-en_GB-common to atrium-api-infix-en_GB-common
  - introduce a function in the companion object for each assertion function so that we see the infix API in action (see CollectionFeatureAssertionsSpec in atrium-api-cc-infix-en_GB-jvm but use fun1, fun2 etc. to retrieve the name of the function instead of `KFunction<...>` -- use `fun1<..>(...).name`)

## Your first contribution?
- Write a comment `I'll work on this` if you would like to take this issue over. 
  This way we get the chance to revise the description in case things have changed in the meantime,
  we might give you additional hints and we can assign the task to you, so that others do not start as well.
- See [Your first code contribution](https://github.com/robstoll/atrium/blob/master/.github/CONTRIBUTING.md#your-first-code-contribution) for guidelines.  
- Do not hesitate to ask questions here or to contact us via [Atrium's slack channel](https://kotlinlang.slack.com/team/U3DE1TXKP) if you need help
  ([Invite yourself](https://slack.kotlinlang.org/) in case you do not have an account yet).

**Comment by ratkayandras** [comment]
I want to work on this.

**Comment by ratkayandras** [comment]
Hi, I don't really know why but I do not get an invitation for the slack channel so I have to ask my questions here.

I was able to create the infix functions in collectionAssertions.kt like:
 `infix fun <T : Collection<*>> Assert<T>.toBe(@Suppress("UNUSED_PARAMETER") Empty: Empty) 
= addAssertion(ExpectImpl.collection.isEmpty(this))`

I was able to create the functions for each assertion function in CollectionAssertionsSpec like:
```
companion object {
       // private fun hasSize(plant: Assert<Collection<Int>>, size: Int): Assert<Collection<Int>>
       //     = plant hasSize size

        private fun isEmpty(plant: Assert<Collection<Int>>)
            = plant toBe Empty

        private fun isNotEmpty(plant: Assert<Collection<Int>>)
            = plant notToBe Empty
    }
```

But I don't understand how should I modify the constructor of CollectionAssertionsSpec
```
class CollectionAssertionsSpec : ch.tutteli.atrium.specs.integration.CollectionAssertionsSpec(
    fun0(Expect<Collection<Int>>::isEmpty),
    fun0(Expect<Collection<Int>>::isNotEmpty)
)
```

I can't use fun1 or fun2 in it I think I am missing a step to fully understand what should I do with this. Any help would be much appreciated.

**Comment by robstoll** [comment]
I guess you have tried: https://surveys.jetbrains.com/s3/kotlin-slack-sign-up
I can send you an invitation if you give me your email address (send me an email to rstoll@tutteli.ch if you don't want to post it here).

Concerning your question. You can use `fun0`, `fun1` etc. this should work. However, you don't have the same signature as in fluent and your function name is different. Which means, instead of `fun0(Expect<Collection<Int>>::isEmpty)` you have to write `fun1(Expect<Collection<Int>>::toBe)` and most likely you have to provide the type parameters (as the Kotiln type inference system is not good enough -- intellij might tell you something different but most likely this is wrong and gradle would fail):  `fun1<Collection<Int>, Empty>(Expect<Collection<Int>>::toBe)` 

However, the goal is not to just simply pass `fun1..` to `ch.tutteli.atrium.specs.integration.CollectionAssertionsSpec(...)` as this does not prove that you have setup the infix API correctly. Thus you should do the following:
```
fun1<Collection<Int>, Empty>(Expect<Collection<Int>>::toBe).name to Companion:isEmpty
```
and in the companion object of the Spec you write:
```
fun isEmpty(expect: Assert<Collection<Int>>) = expect toBe Empty
```
This proves that you can use the API in an infix style. 

I hope this helps

**Comment by robstoll** [comment]
Please do not add an `hasSize` function, this one is will be deprecated on the old infix API and should not be included in the new infix API. Why did you want to add it, is it somewhere in the fluent API? (I am asking because this would be a mistake)

The CollectionFeatureAssertionsSpec shall test the property `size` which is not infix, so for this you can use `property<Collection<String>, Int>(Expect<Collection<String>>::size),` as you already did but for the second argument you should pass `size` which expects an `assertionCreator`

**Comment by ratkayandras** [comment]
Okay thank you for your answer.
```
/**
 * Expects that the subject of the assertion (a [Collection]) has the given [expected] size.
 *
 * Shortcut for `expected.toBe(expectedSize)`.
 *
 * @return This assertion container to support a fluent API.
 * @throws AssertionError Might throw an [AssertionError] if the assertion made is not correct.
 */
fun <T : Collection<*>> Expect<T>.hasSize(expected: Int) = size { toBe(expected) }
```
This is in `package ch.tutteli.atrium.api.fluent.en_GB` and this is why I assumed that you want this function in the infix api as well.

**Comment by robstoll** [comment]
Ok, keep it for now, I can still remove it later on, maybe I changed my mind and I cannot remember :older_man:

**Comment by robstoll** [comment]
Implemented by 62ff3ff3833095c7981e42e98716959c267f0569
